x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,359 @@
+                                    + TextExpansionResults.NAME
+                                    + "] received ["
+                                    + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                    + "]. Is ["
+                                    + inferenceId
+                                    + "] a compatible model?"
+                            )
+                        );
+                    }
+                }, listener::onFailure)
+            )
+        );
+
+        return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+    }
+
+    @Override
+    protected boolean doEquals(SparseVectorQueryBuilder other) {
+        return Objects.equals(fieldName, other.fieldName)
+            && Objects.equals(tokenPruningConfig, other.tokenPruningConfig)
+            && Objects.equals(queryVectors, other.queryVectors);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(fieldName, queryVectors, tokenPruningConfig);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SPARSE_VECTOR_QUERY_ADDED;
+    }
+
+    private static final ConstructingObjectParser<SparseVectorQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME, a -> {
+        String fieldName = (String) a[0];
+        @SuppressWarnings("unchecked")
+        List<WeightedToken> weightedTokens = parseWeightedTokens((Map<String, Object>) a[1]);
+        String inferenceId = (String) a[2];
+        String text = (String) a[3];
+        Boolean shouldPruneTokens = (Boolean) a[4];
+        TokenPruningConfig tokenPruningConfig = (TokenPruningConfig) a[5];
+        return new SparseVectorQueryBuilder(fieldName, weightedTokens, inferenceId, text, shouldPruneTokens, tokenPruningConfig);
+    });
+
+    private static List<WeightedToken> parseWeightedTokens(Map<String, Object> weightedTokenMap) {
+        List<WeightedToken> weightedTokens = null;
+        if (weightedTokenMap != null) {
+            weightedTokens = new ArrayList<>();
+            for (Map.Entry<String, Object> entry : weightedTokenMap.entrySet()) {
+                String token = entry.getKey();
+                Object weight = entry.getValue();
+                if (weight instanceof Float) {
+                    WeightedToken weightedToken = new WeightedToken(token, (Float) entry.getValue());
+                    weightedTokens.add(weightedToken);
+                } else if (weight instanceof Double) {
+                    WeightedToken weightedToken = new WeightedToken(token, ((Double) entry.getValue()).floatValue());
+                    weightedTokens.add(weightedToken);
+                } else {
+                    throw new IllegalArgumentException("Weight must be a float or double");
+                }
```

Comment:
```suggestion
                if (weight instanceof Number number) {
                    WeightedToken weightedToken = new WeightedToken(token, number.floatValue());
                    weightedTokens.add(weightedToken);
                } else {
                    throw new IllegalArgumentException("Weight must be a number");
                }
```

Seems like we should allow any `Numeric` field and grab `floatValue`

Reply:
Much cleaner, thanks 🙏